File: Problem010-Regular Expression Matching/Solution.cpp
/*
 * @SEU(master);@ZZULI(bachelor)
 * author:CuiXuan
 * email:873059043@qq.com
 * If you find any errors, please let me know.
 * If you have any better solution, please let me know.
 *
 * Solution Using time 407ms
 */
class Solution {
public:
    bool isMatch(string s, string p) {
        return Match(s,0,p,0);
    }
    bool Match(string s,int index1,string p,int index2){
        if(index2>=p.size())return index1>=s.size();
        if(index2+1<p.size() && p[index2+1]=='*'){
            while(index1<s.size() && (p[index2]=='.' || p[index2]==s[index1])){
                if(Match(s,index1,p,index2+2))return true;
                index1++;
            }
            return Match(s,index1,p,index2+2);
        }else if(index1<s.size() && (p[index2]=='.' || p[index2]==s[index1])){
            return Match(s,index1+1,p,index2+1);
        }
        return false;
    }
};

File: Problem010-Regular Expression Matching/Solution.java
/*
 * @SEU(master);@ZZULI(bachelor)
 * author:CuiXuan
 * email:873059043@qq.com
 * If you find any errors, please let me know.
 * If you have any better solution, please let me know.
 * 
 * Solution Using time 302ms
 */
public class Solution {
    public boolean isMatch(String s, String p) {
        char[] chs = s.toCharArray();
        char[] chp = p.toCharArray();
        return Match(chs,0,chp,0);
    }
    public boolean Match(char[] chs,int index1,char[] chp,int index2){
        if(index2>=chp.length)return index1>=chs.length;
        if(index2+1<chp.length && chp[index2+1]=='*'){
            while(index1<chs.length && (chp[index2]=='.' || chp[index2]==chs[index1])){
                if(Match(chs,index1,chp,index2+2))return true;
                index1++;
            }
            return Match(chs,index1,chp,index2+2);
        }else if(index1<chs.length && (chp[index2]=='.' || chs[index1]==chp[index2])){
            return Match(chs,index1+1,chp,index2+1);
        }
        return false;
    }
}